src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1571,32 +960,87 @@ where
     ((ok_streams, err_stream), None)
 }
 
+/// Reclocks an `IntoTime` frontier stream into a `FromTime` frontier stream. This is used for the
+/// feedback edge that converts the `IntoTime` resumption frontier into the `FromTime` frontier
+/// that is used with the source's `OffsetCommiter`.
+fn reclock_resume_upper<FromTime, IntoTime>(
+    mut resume_rx: UnboundedReceiver<EventCore<IntoTime, Vec<()>>>,
+    mut reclock_follower: ReclockFollower<FromTime, IntoTime>,
+    worker_id: usize,
+    id: GlobalId,
+) -> impl futures::stream::Stream<Item = Antichain<FromTime>>
+where
+    FromTime: SourceTimestamp,
+    IntoTime: Timestamp + Lattice + Display,
+{
+    async_stream::stream!({
+        let mut resume_upper = MutableAntichain::new_bottom(Timestamp::minimum());
+        let mut prev_upper = reclock_follower.since().to_owned();
+
+        while let Some(event) = resume_rx.recv().await {
+            if let EventCore::Progress(changes) = event {
+                resume_upper.update_iter(changes);
+
+                let source_upper = loop {
+                    match reclock_follower.source_upper_at_frontier(resume_upper.frontier()) {
+                        Ok(frontier) => break frontier,
+                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
+                            tokio::time::sleep(Duration::from_millis(100)).await
+                        }
+                        Err(err) => panic!("unexpected reclock error {:?}", err),
+                    }
+                };
+
+                debug!(
+                    "timely-{worker_id} source({id}) converted resume upper: into_upper={} source_upper={}",
+                    resume_upper.pretty(), source_upper.pretty()
+                );
+
+                // In order to *invert* a frontier the since frontier must be strictly less than
+                // the frontier we're inverting. This means we can't compact all the way to
+                // `resume_upper`, since that would make it potentially equal and lead to wrong
+                // results. Since this is a generic method we can't do the "time minus one" thing
+                // we usually do. So instead we keep track of the previous since frontier and we
+                // use this lagging frontier to compact our trace.
+                if PartialOrder::less_than(&prev_upper.borrow(), &resume_upper.frontier()) {
```

Comment:
I think this is fine? But as a followup, lets consider whether or not we should add the subtraction as a bound on this! Also, I think this now makes more sense?

```suggestion
                // In order to *compact* to a since frontier, it must be strictly not beyond
                // the frontier we're inverting. This means we can't compact all the way to
                // `resume_upper`, since that would make it potentially equal and lead to wrong
                // results. Since this is a generic method we can't do the "time minus one" thing
                // we usually do. So instead we keep track of the previous since frontier and we
                // use this lagging frontier to compact our trace.
                if PartialOrder::less_than(&prev_upper.borrow(), &resume_upper.frontier()) {
```

Reply:
I think the original wording is correct. This was referring to the `since` frontier of the remap collection, not that we're trying to `invert` a `since` frontier